substra/sdk/client.py

```diff
@@ -949,3 +951,111 @@ def organization_info(self) -> models.OrganizationInfo:
+
+        Returns:
+            models.ComputePlan: the compute plan after completion
+        """
+        asset_getter = self.get_compute_plan
+        status_failed = models.ComputePlanStatus.failed.value
+        status_canceled = models.ComputePlanStatus.canceled.value
+        statuses_stopped = (
+            models.ComputePlanStatus.done.value,
+            models.ComputePlanStatus.failed.value,
+            models.ComputePlanStatus.canceled.value,
+        )
+
+        return self._wait(
+            key=key,
+            asset_getter=asset_getter,
+            polling_period=polling_period,
+            raises=raises,
+            status_canceled=status_canceled,
+            status_failed=status_failed,
+            statuses_stopped=statuses_stopped,
+            timeout=timeout,
+        )
+
+    @logit
+    def wait_task(
+        self, key: str, *, timeout: Optional[float] = None, polling_period: float = 2.0, raises: bool = True
+    ) -> models.Task:
+        """Block the execution until the task finishes.
+
+        It is considered finished when the status is done, failed or cancelled.
+        If a timeout is defined, will raises a `FutureTimeoutError` once it reached it.
+
+        Args:
+            key (str): the key of the task to wait
+            timeout (Optional[float]): maximum time to wait. If set to None, will hang until completion.
+            polling_period (float): time to wait between to checks, in seconds. Default to 2.0.
+            raises (bool): wether or not to raises exception if the execution fails. Default to True.
+
+        Returns:
+            models.Task: the task after completion
+        """
+        asset_getter = self.get_task
+        status_canceled = models.Status.canceled.value
+        status_failed = models.Status.failed.value
+        statuses_stopped = (models.Status.done.value, models.Status.canceled.value)
+        return self._wait(
+            key=key,
+            asset_getter=asset_getter,
+            polling_period=polling_period,
+            raises=raises,
+            status_canceled=status_canceled,
+            status_failed=status_failed,
+            statuses_stopped=statuses_stopped,
+            timeout=timeout,
+        )
+
+    def _wait(
+        self,
+        *,
+        key: str,
+        asset_getter,
+        polling_period: float,
+        raises: bool,
+        status_failed: str,
+        status_canceled: str,
+        statuses_stopped: Sequence[str],
+        timeout: Optional[float] = None,
+    ):
+        tstart = time.time()
+        while True:
+            asset = asset_getter(key)
+
+            if asset.status in statuses_stopped:
+                break
+
+            if asset.status == models.Status.failed.value and asset.error_type is not None:
+                # when dealing with a failed task, wait for the error_type field of the task to be set
+                # i.e. wait for the registration of the failure report
+                break
```

Comment:
I'm not sure of the expected behaviour here. Shouldn't we raise the asset.error_type ? Like breaking the loop here will just stop without a message right ?

Reply:
That's a special case for `models.Task` : When you call `_wait` with a `models.Task`, `asset_failed = models.Status.failed.value`but this value is not included in `asset_stopped`. This would wait the field `error_type` to be filled before breaking the loop, and then as `asset.status == status_failed` (`== models.Status.failed.value`) it will `raise exceptions.FutureFailureError(f"Future execution failed on {asset}")`, with `asset` being displayed as a dict with all its content. But we can change that if you think we should just return a more tailored error.